### PR TITLE
Implemented struct constructors

### PR DESCRIPTION
**Pull request by Weltspear**
Hello again.
Implemented Java like struct constructors which work like that:
```rust
import "std/memory.cup"

struct Foo{
	i: i32;
}

Foo(a: i32){
	let ptr: Foo* = malloc(sizeof(Foo));
	ptr.i = a;
	return ptr;
}

fn main(){
	let a: Foo* = new Foo(5);
	print(a.i);
}
```

### CODE REVIEW

**Comment by Weltspear** [comment]
Those constructors work in pretty simple way. Essentially when you create a constructor a function called `new_A` is created, it is marked as a constructor, when constructor is called parser checks whether function `new_A` is a constructor and if it is one the constructor is being called. You can also explicitly call `new_A` function.

**Comment by Weltspear** [comment]
*Constructors also work for `union`s

**Comment by mustafaquraish** [comment]
some nice work here! However, this feels a little hacky. Here's some comments:
 
1. I don't like how the declaration of the constructor looks. All other functions/methods/structs etc have an "identifying" keyword in front of it, which makes parsing nicer. We can probably have this be a method as well:

```rust
method Foo(a: i32) { ... }
``` 

2. The manual allocation here isn't great as well. I would expect the existence of the `new` keyword to do the allocation. Here's a suggestion: if you implement (1), methods implicitly take in a `this` pointer currently, so we can automatically allocate space of the right size and pass that in as the `this` value (we know the size of the struct at compile time).

3. (This idea doesn't need to be part of this PR, but leaving it here anyway) If we're now implicitly passing a pointer to a struct for the constructor to initialize, there's no reason we can't also use it to initialize structs on the stack! Here's what I have in my mind:

```rust
// With allocation
let x = new Foo(5); 
// translates internally to:
let x: Foo* = malloc(sizeof(Foo));
new_Foo(x, 5); 

// On the stack
let x = Foo(5)
// translates (roughly) to:
let x: Foo; 
new_Foo(&x, 5);
```

4. I don't like that we're essentially auto-generating a function name (which the user can create manually). I don't like the idea of "magic methods" which have some syntax over them. I think a nicer way to go about this would be to just change how constructors are named in `get_func_name()` (located in `codegen.cup`). Basically, I don't want the user to be able to call the constructor function without going through the proposed syntax; it's nice to have *one* way of doing something.

I think these changes would make parsing nicer, have easier semantics for the user, and also work for both stack-heap allocation (based on the existence of `new`).

Thoughts?

**Comment by Weltspear** [comment]
1. I don't think constructor needs an indentifying keyword because it is not hard to parse a constructor even without a keyword. Plus I don't think constructor should be treated as a method. If we really want a keyword I would give it `constructor` keyword or something like that idk. Also we could go further and create `impl` blocks like those in rust to separate methods, constructors etc. for them to have their own space.
2. I can make `new` allocate memory for the struct but the question is whether it is gonna require the end user to `import` `std/memory.cup` or not. Instead of `malloc`ing memory we can make `new` pass the pointer to `char` array which has the size of the struct which is on the stack.   
3. We can do this later ig.
4. I'll take a look at that.

Edit: I came up with an idea we can basically create a statement which would take an allocator function to declare default allocator, it would look something like this `__NEW_DEFAULT_ALLOCATOR <func>` and when constructor is called it would call this function to allocate memory for the struct.

**Comment by Weltspear** [comment]
Also about those `impl` blocks we could do something like that:
```rust
impl A{
	A(i: i32){
		...
	}
	
	// pass an explicit self so we could create static methods 
	A::a(self*, i: i32){
		...
	}
	
	// static method
	A::b(){
		...
	}
}
```
What do you think?

**Comment by Weltspear** [comment]
Right so in order to give more flexibility I added `cdefault` statement. Essentially it is used to set default compilation parameters. The only parameter currently is default allocator for `new`. You can specify allocator by writing `cdefault allocator <func>`.

**Comment by Weltspear** [comment]
After some work, now they work like that:
```rust
import "std/memory.cup"

struct A{
	i: i32;
}

cdefault allocator malloc;

A(a: i32){
        // No need for malloc
	self.i = a;
        // Implicit return
}

fn main(){
	let a: A* = new A(5); 
}
```

**Comment by mustafaquraish** [comment]
> 1. I don't think constructor needs an indentifying keyword b

It doesn't _need_ one, but I would prefer having one for it to be syntactically consistent with other stuff. It doesn't need to be `method`, we can do something like `constuctor Foo(...) {...}`

> `impl` blocks

Once again, not a fan of these. They indent all code by 1 level (unnecessarily IMO), which is why I have also opted to leave method definitions outside the class. I don't think they add anything valuable, especially since we don't have any sort of traits or whatever.

> whether it is gonna require the end user to import `std/memory.cup`

I think this is a fair enough assumption to make for now. At some point I want to make it so that we automatically import `std/common.cup` anyway. Having to write imports for such basic tasks is feeling a bit tedious.

> 3. We can do this later ig.

cool.

> I added cdefault statement

I don't think this is particularly necessary. This is mostly an educational project, and will likely never need the flexibility of using different types of allocators for constructing new objects. I prefer only adding stuff to the language when we see an actual benefit from it, so we're not maintaining unused code.

> no need for malloc, implicit return.

Very cool! I'm liking how these constructors are looking.

**Comment by Weltspear** [comment]
> > 1. I don't think constructor needs an indentifying keyword b
> 
> It doesn't _need_ one, but I would prefer having one for it to be syntactically consistent with other stuff. It doesn't need to be `method`, we can do something like `constuctor Foo(...) {...}`
> 
> > `impl` blocks
> 
> Once again, not a fan of these. They indent all code by 1 level (unnecessarily IMO), which is why I have also opted to leave method definitions outside the class. I don't think they add anything valuable, especially since we don't have any sort of traits or whatever.
> 
> > whether it is gonna require the end user to import `std/memory.cup`
> 
> I think this is a fair enough assumption to make for now. At some point I want to make it so that we automatically import `std/common.cup` anyway. Having to write imports for such basic tasks is feeling a bit tedious.
> 
> > 3. We can do this later ig.
> 
> cool.
> 
> > I added cdefault statement
> 
> I don't think this is particularly necessary. This is mostly an educational project, and will likely never need the flexibility of using different types of allocators for constructing new objects. I prefer only adding stuff to the language when we see an actual benefit from it, so we're not maintaining unused code.
> 
> > no need for malloc, implicit return.
> 
> Very cool! I'm liking how these constructors are looking.

Also if you want to use `return` in constructors you just do `return;` instead of `return self;`

**Comment by Weltspear** [comment]
Can we keep `cdefault` for internal use? We can put one for `malloc` into `std/memory.cup` and don't document it ig.

**Comment by Weltspear** [comment]
I chose `_init` instead of `init` because `init` was used in compiler a lot and I didn't want to refactor.

**Comment by Weltspear** [comment]
```rust
import "std/memory.cup"

struct A{
	i: i32;
}

cdefault allocator malloc;

_init A(b: i32){
	self.i = b;
}

fn main(){
	let a: A* = new A(5);
	let b = 5;
	
	print(a.i);
}

```
Current state of constructors ig.

**Comment by mustafaquraish** [comment]
> Can we keep `cdefault` for internal use? We can put one for `malloc` into `std/memory.cup` and don't document it ig.

I guess we can, but it seems rather pointless to do so. I think it's nicer to just assume that `std/memory.cup` has been imported, and we can call `find_function_definition("malloc", ...)` inside the compiler to get a reference to it. If you have an actual use-case for it having this be configurable inside the actual cup code, then sure.

> I chose `_init` instead of `init`

`_init` looks  very ugly. Can we just stick with `constructor` or `initialize`? I don't mind it being a bit verbose. Sorry for being picky here, but I've tried very hard to avoid the quirky syntax some other languages use and make stuff look as consistent as possible, and `_init` just stands out like a sore thumb.

**Comment by mustafaquraish** [comment]
Here's a thought, why not add a command line flag that lets you optionally specify the allocator, and have it default to `malloc`? So something like

```
./cupcc file.cup --allocator debug_malloc
```

**Comment by Weltspear** [comment]
Hmmmh that's an interesting idea.

**Comment by Weltspear** [comment]
Can we implement this parameter in a separate PR?

**Comment by Weltspear** [comment]
`constructor` was used in compiler too so we are gonna use this one Ig

**Comment by mustafaquraish** [comment]
> `constructor` was used in compiler too so we are gonna use this one Ig

That's fine. Should be trivial enough to change the name later if we need to. This implementation looks good to me for now, would you just mind adding some tests for it as well?

**Comment by Weltspear** [comment]
wait a sec

**Comment by Weltspear** [comment]
I'm gonna add that paramater

**Comment by Weltspear** [comment]
to the compiler

**Comment by Weltspear** [comment]
Indentation is a bit messed up in some files.

**Comment by Weltspear** [comment]
Alright, now please do your testing magic Ig.

**Comment by mustafaquraish** [comment]
> Alright, now please do your testing magic Ig.

You can see how tests are written in the `tests/` folder. Essentially, it's a shell script comprising of code snippets, and for each code snippet you can use the utilities defined to either (1) expect it to exit with a certain value or (2) expect it to output certain text.

Here's a file that contains multiple ways of doing so, you should be able to generalize: https://github.com/mustafaquraish/cup/blob/master/tests/core.sh

I think an appropriate for you to add some tests would be in `tests/types.sh`.

**Comment by Weltspear** [comment]
Okay

File: compiler/main.cup [hunk 1]
```diff
@@ -47,7 +49,11 @@ fn parse_cli_args(argc: int, argv: char**): char** {
         } else if (streq(argv[i], "-o")) {
             ++i;
             m_output_filename = argv[i];
-        } else if (m_filename == null) {
+        } else if (streq(argv[i], "--alloc")){
+			++i;
+			p_default_allocator = malloc(strlen(argv[i])*sizeof(char));
+			strcpy(p_default_allocator, argv[i]);
```
Comment (mustafaquraish) [inline compiler/main.cup]: allocation here is unnecessary. You can just do `p_default_allocator = argv[i]`

File: compiler/parser.cup [hunk 2]
```diff
@@ -461,27 +461,31 @@ fn parse_factor(lexer: Lexer*): Node* {
 		if (func != null){	
 			// Getting default allocator
 			if (p_default_allocator == null){
-				die_loc(here, &struct_name.loc, "Default memory allocator hasn't been defined. Import std/memory.cup for a default allocator.");
+				die_loc(here, &struct_name.loc, "Memory allocator hasn't been defined.");
 			}
```
Comment (mustafaquraish) [inline compiler/parser.cup]: This should never happen now

File: compiler/parser.cup [hunk 3]
```diff
@@ -461,27 +461,31 @@ fn parse_factor(lexer: Lexer*): Node* {
...
-				die_loc(here, &struct_name.loc, "Default allocator doesn't have a correct signature. It should take an i32 as a first arg and return a void pointer.");
+				die_loc(here, &struct_name.loc, "Allocator doesn't have a correct signature. It should take an i32 as a first arg and return a void pointer.");
 			}
 			
 			if (allocator_func.etyp.typ != TYPE_PTR){
-				die_loc(here, &struct_name.loc, "Default allocator doesn't have a correct signature. It should take an i32 as a first arg and return a void pointer.");
+				die_loc(here, &struct_name.loc, "Allocator doesn't have a correct signature. It should take an i32 as a first arg and return a void pointer.");
 			}
 			
 			if (allocator_func.etyp.ptr.typ != TYPE_VOID){
```
Comment (mustafaquraish) [inline compiler/parser.cup]: Since the error message here is the same, perhaps we can collapse it into one condition?

File: compiler/types.cup [hunk 4]
```diff
@@ -35,6 +35,7 @@ struct Type {
 
     fields: Vector*; // Vector<Variable*>;
     methods: Vector*; // Vector<Node(AST_FUNC)*>;
+    constructor: void*; // Node(AST_FUNC)*
```
Comment (mustafaquraish) [inline compiler/types.cup:38]: We can make the type of this `Node *` right?

Comment (Weltspear) [reply compiler/types.cup:38]: We can't make the type `Node*` because that would cause circular dependency

Comment (mustafaquraish) [reply compiler/types.cup:38]: Ah I see. Fair enough.

File: compiler/parser.cup (hunk 3 above)
Comment (Weltspear) [reply compiler/parser.cup]: Done.

File: compiler/parser.cup (hunk 2 above)
Comment (Weltspear) [reply compiler/parser.cup]: Done

File: compiler/main.cup (hunk 1 above)
Comment (Weltspear) [reply compiler/main.cup]: Done

**Comment by Weltspear** [comment]
Can I somehow compile the compiler for MacOS on wsl?

**Comment by Weltspear** [comment]
> some nice work here! However, this feels a little hacky. Here's some comments:
> 
> 1. I don't like how the declaration of the constructor looks. All other functions/methods/structs etc have an "identifying" keyword in front of it, which makes parsing nicer. We can probably have this be a method as well:
> 
> ```rust
> method Foo(a: i32) { ... }
> ```
> 
> 2. The manual allocation here isn't great as well. I would expect the existence of the `new` keyword to do the allocation. Here's a suggestion: if you implement (1), methods implicitly take in a `this` pointer currently, so we can automatically allocate space of the right size and pass that in as the `this` value (we know the size of the struct at compile time).
> 3. (This idea doesn't need to be part of this PR, but leaving it here anyway) If we're now implicitly passing a pointer to a struct for the constructor to initialize, there's no reason we can't also use it to initialize structs on the stack! Here's what I have in my mind:
> 
> ```rust
> // With allocation
> let x = new Foo(5); 
> // translates internally to:
> let x: Foo* = malloc(sizeof(Foo));
> new_Foo(x, 5); 
> 
> // On the stack
> let x = Foo(5)
> // translates (roughly) to:
> let x: Foo; 
> new_Foo(&x, 5);
> ```
> 
> 4. I don't like that we're essentially auto-generating a function name (which the user can create manually). I don't like the idea of "magic methods" which have some syntax over them. I think a nicer way to go about this would be to just change how constructors are named in `get_func_name()` (located in `codegen.cup`). Basically, I don't want the user to be able to call the constructor function without going through the proposed syntax; it's nice to have _one_ way of doing something.
> 
> I think these changes would make parsing nicer, have easier semantics for the user, and also work for both stack-heap allocation (based on the existence of `new`).
> 
> Thoughts?

Btw I came up with an idea, we do can stack allocation with `let a: Foo(...);` syntax, it would be easier to parse if we did it this way.  
Edit: or maybe we can do it with `new` I'll take a look at how structs are built Ig

**Comment by mustafaquraish** [comment]
> `let a: Foo(...)`

I don't particularly like this. To me at least stack allocation should be symmetric to allocation on the heap, just without `new`. 

This syntax wouldn't allow me to re-initialize an existing struct either, no?

**Comment by Weltspear** [comment]
Ye

**Comment by mustafaquraish** [comment]
Anyway, let's leave the stack-constructors for another PR, this one is already decently big. Let's try to use it for a bit and see how it feels, and then we can iterate.

Is there anything else you'd like to add? I'll go ahead and do a final review of everything later today.

Re: macOS testing, you should just be able to push changes to the PR and look at CI results. I don't know of any other way to test macOS through another OS barring a VM.

**Comment by Weltspear** [comment]
> Anyway, let's leave the stack-constructors for another PR, this one is already decently big. Let's try to use it for a bit and see how it feels, and then we can iterate.
> 
> Is there anything else you'd like to add? I'll go ahead and do a final review of everything later today.
> 
> Re: macOS testing, you should just be able to push changes to the PR and look at CI results. I don't know of any other way to test macOS through another OS barring a VM.

I think I have added everything what I wanted to add related to constructors. Btw git question when you're gonna merge how can I make my fork stop showing that I have x commits ahead?

**Comment by mustafaquraish** [comment]
> I think I have added everything what I wanted to add related to constructors. Btw git question when you're gonna merge how can I make my fork stop showing that I have x commits ahead?

Usually in my workflow, I have my fork and the upstream repository (in this case _my_ one) set as remotes. You should then be able to update your code to the upstream version after a merge by `git pull upstream master` I believe. Generally I recommend developing on a new branch and making PRs from that, that way you can always just pull upstream and make a new branch from that to stay up-to-date.

**Comment by Weltspear** [comment]
Alright thx
